File: backend/app/models/learning.py
from datetime import date
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, Date, ForeignKey, Integer, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.base import TimestampMixin

if TYPE_CHECKING:
    from app.models.user import MemberProfile


class Skill(Base, TimestampMixin):
    __tablename__ = "skills"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    category: Mapped[str] = mapped_column(String(50), default="general", index=True, nullable=False)
    description: Mapped[str | None] = mapped_column(String(500), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    member_skills: Mapped[list["MemberSkill"]] = relationship(back_populates="skill")


class MemberSkill(Base, TimestampMixin):
    __tablename__ = "member_skills"
    __table_args__ = (UniqueConstraint("member_id", "skill_id", name="uq_member_skill"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    member_id: Mapped[int] = mapped_column(
        ForeignKey("member_profiles.id", ondelete="CASCADE"), index=True, nullable=False
    )
    skill_id: Mapped[int] = mapped_column(
        ForeignKey("skills.id", ondelete="CASCADE"), index=True, nullable=False
    )
    level: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    note: Mapped[str | None] = mapped_column(String(500), nullable=True)

    member: Mapped["MemberProfile"] = relationship(viewonly=True)
    skill: Mapped["Skill"] = relationship(back_populates="member_skills")


class LearningPlan(Base, TimestampMixin):
    __tablename__ = "learning_plans"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    member_id: Mapped[int] = mapped_column(
        ForeignKey("member_profiles.id", ondelete="CASCADE"), index=True, nullable=False
    )
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    category: Mapped[str] = mapped_column(String(50), default="general", index=True, nullable=False)
    start_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    target_date: Mapped[date | None] = mapped_column(Date, nullable=True, index=True)
    status: Mapped[str] = mapped_column(String(20), default="not_started", index=True, nullable=False)
    progress: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_by: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )

File: frontend/src/api/learning.ts
import client from './client'

export interface Skill {
  id: number
  name: string
  category: string
  description: string | null
  sort_order: number
  is_active: boolean
}

export interface MemberSkill {
  id: number
  skill_id: number
  level: number
  note: string | null
  updated_at: string
}

export interface LearningPlan {
  id: number
  member_id: number
  title: string
  description: string | null
  category: string
  start_date: string | null
  target_date: string | null
  status: string
  progress: number
  created_by: number | null
  created_at: string
  updated_at: string
}

export function listSkills() {
  return client.get<{ data: Skill[] }>('/skills').then((r) => r.data.data)
}

export function createSkill(payload: Record<string, unknown>) {
  return client.post<{ data: Skill }>('/skills', payload)
}

export function getMemberSkills(memberId: number) {
  return client
    .get<{ data: MemberSkill[] }>(`/members/${memberId}/skills`)
    .then((r) => r.data.data)
}

export function setMemberSkills(memberId: number, items: { skill_id: number; level: number; note?: string }[]) {
  return client.put<{ data: MemberSkill[] }>(`/members/${memberId}/skills`, items)
}

export function listPlans(params: { member_id?: number; status?: string } = {}) {
  return client.get<{ data: LearningPlan[] }>('/learning-plans', { params })
}

export function createPlan(payload: Record<string, unknown>) {
  return client.post<{ data: LearningPlan }>('/learning-plans', payload)
}

export function updatePlan(id: number, payload: Record<string, unknown>) {
  return client.patch<{ data: LearningPlan }>(`/learning-plans/${id}`, payload)
}

export function deletePlan(id: number) {
  return client.delete(`/learning-plans/${id}`)
}
